src/lib.rs

```diff
@@ -199,8 +231,81 @@ impl Contract {
         notify::notify_mentions(desc.as_str(), id);
     }
 
+    #[payable]
+    pub fn add_proposal(&mut self, body: VersionedProposalBody, labels: HashSet<String>) {
+        near_sdk::log!("add_proposal");
+        let id = self.proposals.len();
+        let author_id = env::predecessor_account_id();
+        let editor_id = author_id.clone();
+
+        let proposal_body = body.clone().latest_version();
+
+        require!(
+            self.is_allowed_to_use_labels(
+                Some(editor_id.clone()),
+                labels.iter().cloned().collect()
+            ),
+            "Cannot use these labels"
+        );
+        require!(
+            proposal_body.payouts.is_empty(),
+            "Can't add proposal with payouts at the beginning"
+        );
+
+        require!(self.proposal_categories.contains(&proposal_body.category), "Unknown category");
+
+        require!(
+            is_draft(&proposal_body.timeline),
+            "Cannot create proposal which is not in a draft state"
+        );
+
+        for label in &labels {
+            let mut other_proposals = self.label_to_proposals.get(label).unwrap_or_default();
+            other_proposals.insert(id);
+            self.label_to_proposals.insert(label, &other_proposals);
+        }
+
+        let mut author_proposals = self.author_proposals.get(&author_id).unwrap_or_default();
+        author_proposals.insert(id);
+        self.author_proposals.insert(&author_id, &author_proposals);
+
+        let proposal = Proposal {
+            id,
+            author_id: author_id.clone(),
+            social_db_post_block_height: 0u64,
+            snapshot: ProposalSnapshot {
+                editor_id,
+                timestamp: env::block_timestamp(),
+                labels,
+                body: body.clone(),
+            },
+            snapshot_history: vec![],
+        };
+
+        proposal::repost::publish_to_socialdb_feed(
```

Comment:
Return this `Promise` as a result of `add_proposal` call. See https://github.com/NEAR-DevHub/neardevhub-contract/pull/88#discussion_r1476879100 for more details about returning a Promise